iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/PrivilegeEscalationCheck.java

```diff
@@ -19,60 +19,79 @@
 
-  private static final String MESSAGE = "Narrow these permissions to a smaller set of resources to avoid privilege escalation.";
-  private static final Pattern RESOURCE_NAME_PATTERN = Pattern.compile("arn:[^:]*:[^:]*:[^:]*:[^:]*:(role|user|group)/\\*");
+  private static final String MESSAGE = "This policy is vulnerable to the \"%s\" privilege escalation vector. Remove permissions or " +
+    "restrict the set of resources they apply to.";
+  private static final String MESSAGE_ACTION_MULTIPLE = "When combined with others, this permission enables the \"%s\" escalation vector.";
+  private static final String MESSAGE_ACTION_SINGLE = "This permission enables the \"%s\" escalation vector.";
+  private static final String MESSAGE_STATEMENT_ALL = "Permissions are granted on all resources.";
 
   @Override
   protected void registerResourceConsumer() {
     register("aws_iam_policy",
-      resource ->  PolicyUtils.getPolicies(resource.tree)
-        .forEach(policy -> checkPrivilegeEscalation(resource.ctx, policy)));
+      resource -> PolicyUtils.getPolicies(resource.tree)
+        .forEach(policy -> checkPrivilegeEscalation(resource, policy)));
   }
 
-  private static void checkPrivilegeEscalation(CheckContext ctx, Policy policy) {
-    policy.statement().stream()
-      .filter(PrivilegeEscalationCheck::allowsPrivilegeEscalation)
-      .forEach(statement -> ctx.reportIssue(statement.resource().get(), MESSAGE));
-  }
+  private static void checkPrivilegeEscalation(ResourceSymbol resourceSymbol, Policy policy) {
+    for (Statement statement : policy.statement()) {
+      List<Tree> actionTrees = Collections.emptyList();
 
-  private static boolean allowsPrivilegeEscalation(Statement statement) {
-    return statement.effect().filter(PrivilegeEscalationCheck::isAllowEffect).isPresent()
-        && statement.resource().filter(PrivilegeEscalationCheck::isSensitiveResource).isPresent()
-        && statement.action().filter(PrivilegeEscalationCheck::isSensitiveAction).isPresent()
-        && statement.condition().isEmpty()
-        && statement.principal().isEmpty();
-  }
+      Optional<Tree> action = statement.action();
+      if (action.isPresent() && action.get() instanceof TupleTree) {
+        actionTrees = ((TupleTree) action.get()).elements().trees().stream().map(Tree.class::cast).collect(Collectors.toList());
+      }
 
-  private static boolean isAllowEffect(Tree effect) {
-    return TextUtils.isValue(effect, "Allow").isTrue();
+      Optional<PrivilegeEscalationVector> vectorOpt = PrivilegeEscalationVector.getStatementEscalationVector(statement, actionTrees);
+      if (vectorOpt.isPresent()) {
+        PrivilegeEscalationVector vector = vectorOpt.get();
+        String vectorName = vector.getName();
+        List<SecondaryLocation> secondaryLocations = secondaryLocations(statement, vector, vectorName);
+        resourceSymbol.report(String.format(MESSAGE, vectorName), secondaryLocations);
+      }
```

Comment:
What do you think about moving line `63` to `69` in the above `if`. In general, I believe it only makes sense to look for a vector if the above condition is true `action.isPresent() && action.get() instanceof TupleTree`, and then we can avoid declaring `actionTrees` outside to an `emptyList`